storage/series.go

```diff
@@ -333,33 +338,54 @@ func (s *dbSeries) bufferDrained(start time.Time, encoder encoding.Encoder) {
 	// NB(r): this will occur if after bootstrap we have a partial
 	// block and now the buffer is passing the rest of that block
 	stream := encoder.Stream()
-	s.drainStream(s.blocks, stream, timeZero)
+	s.drainStreamWithLock(s.blocks, start, stream)
 	stream.Close()
 }
 
-func (s *dbSeries) drainStream(blocks block.DatabaseSeriesBlocks, stream io.Reader, cutover time.Time) error {
-	iter := s.opts.ReaderIteratorPool().Get()
-	iter.Reset(stream)
+func (s *dbSeries) drainStreamWithLock(
+	blocks block.DatabaseSeriesBlocks,
+	blockStart time.Time,
+	stream xio.SegmentReader,
+) error {
+	readers := make([]io.Reader, 0, 2)
+	if block, ok := blocks.BlockAt(blockStart); ok {
+		ctx := s.opts.ContextPool().Get()
+		defer ctx.Close()
+		reader, err := block.Stream(ctx)
+		if err != nil {
+			return err
+		}
+		readers = append(readers, reader)
+	}
 
-	// Close the iterator and return to pool when done
-	defer iter.Close()
+	readers = append(readers, stream)
 
-	for iter.Next() {
-		dp, unit, annotation := iter.Current()
-		// If the datapoint timestamp is before the cutover, skip it.
-		if dp.Timestamp.Before(cutover) {
-			continue
-		}
-		blockStart := dp.Timestamp.Truncate(s.opts.RetentionOptions().BlockSize())
-		block := blocks.BlockOrAdd(blockStart)
+	multiIter := s.opts.MultiReaderIteratorPool().Get()
+	multiIter.Reset(readers)
+	defer multiIter.Close()
 
-		if err := block.Write(dp.Timestamp, dp.Value, unit, annotation); err != nil {
+	blopts := s.opts.DatabaseBlockOptions()
+	encoder := s.opts.EncoderPool().Get()
```

Comment:
Encoders get left open and it is up to the block to close it when the block is finished with.